Model/Menu/ExportProcessor.php

```diff
@@ -0,0 +1,169 @@
+
+    /**
+     * @var SerializerInterface
+     */
+    private $serializer;
+
+    /**
+     * @var MenuRepositoryInterface
+     */
+    private $menuRepository;
+
+    /**
+     * @var NodeRepositoryInterface
+     */
+    private $nodeRepository;
+
+    /**
+     * @var MenuResource
+     */
+    private $menuResource;
+
+    public function __construct(
+        SearchCriteriaBuilder $searchCriteriaBuilder,
+        Filesystem $filesystem,
+        SerializerInterface $serializer,
+        MenuRepositoryInterface $menuRepository,
+        NodeRepositoryInterface $nodeRepository,
+        MenuResource $menuResource
+    ) {
+        $this->searchCriteriaBuilder = $searchCriteriaBuilder;
+        $this->directory = $filesystem->getDirectoryWrite(DirectoryList::VAR_DIR);
+        $this->serializer = $serializer;
+        $this->menuRepository = $menuRepository;
+        $this->nodeRepository = $nodeRepository;
+        $this->menuResource = $menuResource;
+    }
+
+    /**
+     * @param int $menuId
+     * @return array
+     */
+    public function getExportFileDownloadContent($menuId)
+    {
+        $data = $this->getExportData($menuId);
+        return $this->generateCsvDownloadFile($data[MenuInterface::IDENTIFIER], $data);
+    }
+
+    /**
+     * @param string $fileId
+     * @param array|null $csvHeaders
+     * @return array
+     */
+    public function generateCsvDownloadFile($fileId, array $data, $csvHeaders = null)
+    {
+        $this->directory->create(self::EXPORT_DIR);
+
+        $file = $this->getDownloadFile($fileId);
+        $stream = $this->directory->openFile($file, 'w+');
+        $stream->lock();
+
+        $stream->writeCsv($csvHeaders ?: $this->getMenuFields());
+        $stream->writeCsv($data);
+
+        $stream->unlock();
+        $stream->close();
+
+        return ['type' => 'filename', 'value' => $file, 'rm' => true];
+    }
+
+    /**
+     * @param int $menuId
+     * @return array
+     */
+    private function getExportData($menuId)
+    {
+        $menu = $this->menuRepository->getById($menuId);
+        $stores = $menu->getStores();
+        $data = $menu->getData();
+        $nodes = $this->getMenuNodeList($menuId);
+
+        $data[self::STORES_CSV_FIELD] = implode(',', $stores);
+        $data[self::NODES_CSV_FIELD] = $nodes ? $this->serializer->serialize($nodes) : null;
+
+        unset($data[MenuInterface::MENU_ID]);
+
+        return $data;
+    }
+
+    /**
+     * @param int $menuId
+     * @return array
+     */
+    private function getMenuNodeList($menuId)
+    {
+        $searchCriteria = $this->searchCriteriaBuilder
+            ->addFilter(NodeInterface::MENU_ID, $menuId)
+            ->create();
+
+        $nodes = $this->nodeRepository->getList($searchCriteria)->getItems();
+        $nodesData = [];
+
+        foreach ($nodes as $key => $node) {
+            $nodesData[$key] = $node->getData();
+            unset($nodesData[$key][NodeInterface::MENU_ID]);
```

Comment:
What do you think if we avoid to export node ID as well? I see later, when importing, it's removed, but I think we don't need it in export file either: https://github.com/SnowdogApps/magento2-menu/pull/144/files#diff-64fbf188d4c313434d60faecd6839110c25242e57597f3ca1c7bf073434a9340R125

Reply:
Node ID alongside Parent ID are important in order identify the parent-child relations between nodes.  
(i.e. If a node parent ID matches another node ID, then it's a child of that node.)

I'll apply some changes to the code to make importing node ID and parent ID optional,  
for nodes that don't have children and nodes that don't have parents.

Reply:
OR replace Node IDs with some UUID that are vaild only during export/import to avoid using those in database

OR use some better data format - originally when thinking about this functionality I was envisiong using YAML or JSON so structure can be implied by data format (in theory YAML would be more user friendly allowing manual edditing in almost-human-readable format)